frontend/packages/operator-lifecycle-manager/src/components/create-operand.tsx

```diff
@@ -104,297 +178,465 @@ const defaultValueFor = (field: OperandField) => {
-          if (_.isEmpty(openAPIProperties)) {
-            return { ...field, capabilities };
-          }
-          const schemaPath = field.path.split('.').join('.properties.');
-          const required = (_.get(
-            openAPI?.properties,
-            _.dropRight(['spec', ...field.path.split('.')])
-              .join('.properties.')
-              .concat('.required'),
-            [],
-          ) as string[]).includes(_.last(field.path.split('.')));
-          const type = _.get(openAPIProperties, schemaPath.concat('.type')) as JSONSchema6TypeName;
-          const validation = _.pick(openAPIProperties[schemaPath], [
-            ...Object.keys(Validations),
-          ]) as OperandField['validation'];
-
-          return { ...field, capabilities, type, required, validation };
-        })
-        .concat(
-          fieldsForOpenAPI(openAPI).filter(
-            (crdField) =>
-              providedAPI.specDescriptors &&
-              !providedAPI.specDescriptors.some((d) => d.path === crdField.path),
-          ),
-        )
-        // Associate `specDescriptors` with `fieldGroups` from OpenAPI
-        .map((field, i, allFields) =>
-          allFields.some((f) =>
-            f.capabilities.includes(
-              SpecCapability.fieldGroup.concat(
-                field.path.split('.')[0],
-              ) as SpecCapability.fieldGroup,
-            ),
-          )
-            ? {
-                ...field,
-                capabilities: [
-                  ...new Set(field.capabilities).add(
-                    SpecCapability.fieldGroup.concat(
-                      field.path.split('.')[0],
-                    ) as SpecCapability.fieldGroup,
-                  ),
-                ],
-              }
-            : field,
-        ),
-    [openAPI, providedAPI, clusterServiceVersion],
-  );
+  // Map providedAPI spec descriptors and openAPI spec properties to OperandField[] array
+  const initialFields: OperandField[] = React.useMemo(() => {
+    // Get fields from openAPI
+    const openAPIFields = fieldsForOpenAPI(openAPI, providedAPI, buffer);
+
+    // Get fields from providedAPI that do not exist in the OpenAPI spec.
+    const providedAPIFields = _.reduce(
+      providedAPI?.specDescriptors || [],
+      (providedAPIFieldsAccumulator, specDescriptor) => {
+        // If this field was already created, ignore it.
+        if (_.find(openAPIFields, { path: specDescriptor.path })) {
+          return providedAPIFieldsAccumulator;
+        }
+
+        // Add the field if it doesn't exist
+        return [...providedAPIFieldsAccumulator, ...specDescriptorToFields(specDescriptor, buffer)];
+      },
+      [],
+    );
 
-  const [fields, setFields] = React.useState<OperandField[]>(fieldsFromProps);
+    // Concatenate all fields and return
+    return [...openAPIFields, ...providedAPIFields];
+  }, [openAPI, providedAPI, buffer]);
+  const [fields, setFields] = React.useState<OperandField[]>(initialFields);
```

Comment:
Moved a few initializers, state declarations, and update functions around to group related things together.